internal/blobtesting/eventually_consistent.go

```diff
@@ -0,0 +1,190 @@
+package blobtesting
+
+import (
+	"context"
+	"io/ioutil"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/repo/blob"
+)
+
+const ecCacheDuration = 5 * time.Second
+
+// ecFrontendCache is an instance of cache, which simulates cloud storage frontend
+// with its own in-memory state. This causes eventual consistency when a client uses
+// different instances of the cache for read and writes.
+type ecFrontendCache struct {
+	mu            sync.Mutex
+	cachedEntries map[blob.ID]*ecCacheEntry
+}
+
+func (c *ecFrontendCache) get(id blob.ID) *ecCacheEntry {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	e := c.cachedEntries[id]
+	if e != nil && !e.isValid() {
+		c.sweepLocked()
+		return nil
+	}
+
+	return e
+}
+
+func (c *ecFrontendCache) put(id blob.ID, data []byte) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.sweepLocked()
+
+	if c.cachedEntries == nil {
+		c.cachedEntries = map[blob.ID]*ecCacheEntry{}
+	}
+
+	if data != nil {
+		// clone data before storage
+		data = append([]byte(nil), data...)
+	}
+
+	c.cachedEntries[id] = &ecCacheEntry{
+		accessTime: time.Now(),
+		data:       data,
+	}
+}
+
+func (c *ecFrontendCache) sweepLocked() {
+	for k, v := range c.cachedEntries {
+		if !v.isValid() {
+			delete(c.cachedEntries, k)
+		}
+	}
+}
+
+type ecCacheEntry struct {
+	data []byte
+
+	accessTime time.Time
+}
+
+func (e *ecCacheEntry) isValid() bool {
+	return time.Since(e.accessTime) < ecCacheDuration
+}
+
+type eventuallyConsistentStorage struct {
+	mu sync.Mutex
+
+	listDropProbability float64
+
+	caches      []*ecFrontendCache
+	realStorage blob.Storage
+}
+
+func (s *eventuallyConsistentStorage) randomFrontendCache() *ecFrontendCache {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	n := rand.Intn(len(s.caches))
+
+	if s.caches[n] == nil {
+		s.caches[n] = &ecFrontendCache{}
+	}
+
+	return s.caches[n]
+}
+
+func (s *eventuallyConsistentStorage) GetBlob(ctx context.Context, id blob.ID, offset, length int64) ([]byte, error) {
+	// don't bother caching partial reads
+	if length >= 0 {
+		return s.realStorage.GetBlob(ctx, id, offset, length)
+	}
+
+	c := s.randomFrontendCache()
+
+	// see if the frontend has the blob cached
+	e := c.get(id)
+	if e != nil {
+		if e.data == nil {
+			return nil, blob.ErrBlobNotFound
+		}
+
+		return append([]byte(nil), e.data...), nil
+	}
+
+	// fetch from the underlying storage.
+	v, err := s.realStorage.GetBlob(ctx, id, offset, length)
+	if err != nil {
+		if err == blob.ErrBlobNotFound {
+			c.put(id, nil)
+		}
+
+		return nil, err
+	}
+
+	c.put(id, v)
+
+	return v, nil
+}
+
+func (s *eventuallyConsistentStorage) PutBlob(ctx context.Context, id blob.ID, data blob.Bytes) error {
+	if err := s.realStorage.PutBlob(ctx, id, data); err != nil {
+		return err
+	}
+
+	d, err := ioutil.ReadAll(data.Reader())
+	if err != nil {
+		return errors.Wrap(err, "invalid data")
+	}
+
+	// add to frontend cache
+	s.randomFrontendCache().put(id, d)
```

Comment:
Do we have to put it to the frontendCache? The behaviour would be the same even without this line since GetBlob anyway checks the underlying storage in case an object is not found in the front end cache.

Reply:
Okay, I take back the comment. The place where this would make a difference is in DeleteBlob. Nevermind.